# MGE via snakemake with added functionality for BGE #
**Requirements:**
- 2_snakefile-insecta and 2_snakefile-genefetch require samples.csv (e.g. BGE_test_samples.csv) listed in config.yaml files (see below):
  
| ID | forward | reverse | taxid |
| --- | --- | --- | --- |
| BSNHM002-24  | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 177658 |
| BSNHM038-24 | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 177627 |
| BSNHM046-24 | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 3084599 |

- 2_snakefile-genefetch require protein_references.csv (e.g. gene_fetch_BGE_test_data.csv) listed in config.yaml (see below):

| ID | matched_term | accession_number | reference_path | reference_name |
| --- | --- | --- | --- | --- |
| BSNHM002-24  | Apataniidae | YP_010586031.1 | abs/path/to/protein_references/BSNHM002-24.fasta | BSNHM002-24 |
| BSNHM038-24 | Trichoptera | YP_010894795.1 | abs/path/to/protein_references/BSNHM038-24.fasta | BSNHM038-24 |
| BSNHM046-24 | Polycentropodidae | YP_010426350.1 | abs/path/to/protein_references/BSNHM046-24.fasta | BSNHM046-24 |
  
- Snakemake, TrimGalore, Exonerate, and numpy installed in conda env
  
- Run using snakemake.sh

**Test run**
- Raw reads for 12 test samples can be downloaded [here](https://naturalhistorymuseum-my.sharepoint.com/personal/b_price_nhm_ac_uk/_layouts/15/onedrive.aspx?ct=1723035606962&or=Teams%2DHL&ga=1&LOF=1&id=%2Fpersonal%2Fb%5Fprice%5Fnhm%5Fac%5Fuk%2FDocuments%2F%5Ftemp%2F%5FBGEexamples4Felix%2F1%5Fraw%5Fdata). Each read pair must be in seperate subdirectories under a parent directory that can be called anything.
- BGE_test_samples.csv and gene_fetch_BGE_test_data.csv listed in config.yaml.

## Running options: ##
**2_snakefile-insecta:**
- Uses config-insecta.yaml
  - Contains path to 'samples.csv'.
  - Contains different parameter configurations for -s and -r.
- Uses insecta_cox1.fasta as protein reference (stored in protein_references dir).

**2_snakefile-genefetch:** 
- Uses config-genefetch.yaml
  - Contains path to 'samples.csv', an output directory, and 'protein_references.csv' (generated by 1_gene_fetch.py)
- Uses taxa-specific references from protein_references.csv (examples stored in protein_references_2 dir).

## To do ##
- Get different parameter configurations to run in conjunction with insecta AND genefetch protein references (see 2_snakefile-insecta for parameter change implementation).
- Integrate 3_mge_tidy-snakemake.py and 4_mge_stats-snakemake.py (funcitonality) into snakefile

## MGE output post-processing ##
Python scripts to organise outputs into directories, trim consensus sequences and 'clean' header names, concatenate into a multi-fasta, calculate summary stats and run BLAST on each consensus sequence.

### 3_mge_tidy-snakemake.py ###
Usage information output by running 'python 3_mge_tidy-snakemake.py'
- Organised files into subdirectories (alignment, consensus, err, logs, out) depending on file extension.
- Concatenates consensus.fas files together (within consensus subdir) and outputs user-specified multi-fasta (highly recommended to include 'cox1' in concatenated consensus filename for downstream BOLDigger2 runs).
- 'Cleans' and trims sequence headers of each consenus sequences.

### 4_mge_stats-snakemake.py ###
Usage information output by running 'python 4_mge_stats-snakemake.py'. Outputs summary stats to .csv file.
- Reads relevant fields from .out files in 'out' dir:
  -  consensus sequence output (i.e. process ID)
  -  Number of input sequences considered
  -  sequences found in vulgar file
  -  number of aligned reads
  -  number of skipped reads due to low rel. score (low relative score = read didn't pass exonerate relative score threshold (i.e. -r))
  -  length of alignment (i.e. reference sequence length)
  -  Coverage minimum
  -  Coverage maximum
  -  Coverage mean
  -  Coverage median
- Determines sequence stats from concatenated consensus.fasta generated by '4_mge_tidy-snakemake.py':
  - Sequence length (only GTACs)
  - N Count
  - Dash and Tilde Count
  - Non-GTAC Count (total of 'N Count' and 'Dash and Tilde Count')
- Parses BLASTn (against 'nt' DB) results:
  - BLAST Title (top hit sequence name)
  - BLAST Score (HSP score assigned to the alignment (higher = better)
  - BLAST alignment length
  - BLAST Identity (number of identical base matches between the query and subject sequences)
  - BLAST percent identity
  - BLAST Gaps
  - BLAST Query Coverage (range of positions in query (input) sequence covered by alignment)
  - BLAST Subject Coverage (range of positions in subject equence covered by alignment)
  - BLAST Accession (top hit accession number)
