# MGE via snakemake with added functionality for BGE #
## Requirements: ##
- snakefile
- config.yaml (containing:)
  - path to'samples.csv' (example below - created via [sample-processing](https://github.com/bge-barcoding/sample-processing) workflow)
  - path to 'protein_references.csv' (example below - created using [1_gene_fetch.py](https://github.com/SchistoDan/MitoGeneExtractor/blob/main/snakemake/1_gene_fetch.py))
  - path to output directory (new directories will be created)
  - gene of interest (e.g. cox1)
  - r (Exonerate relative score threshold) and s (Exonerate minimum score threshold) parameter changes
- Activated conda env (with Snakemake, TrimGalore, Exonerate, Fastp, Biopython and Numpy installed). See mge_env.yaml
- Can be run on cluster using 'snakemake.sh'

**samples.csv example**
| ID | forward | reverse | taxid |
| --- | --- | --- | --- |
| BSNHM002-24  | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 177658 |
| BSNHM038-24 | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 177627 |
| BSNHM046-24 | abs/path/to/R1.fq.gz | abs/path/to/R2.fq.gz | 3084599 |

**protein_references.csv example** 
| ID | matched_term | accession_number | reference_path | reference_name |
| --- | --- | --- | --- | --- |
| BSNHM002-24  | Apataniidae | YP_010586031.1 | abs/path/to/protein_references/BSNHM002-24.fasta | BSNHM002-24 |
| BSNHM038-24 | Trichoptera | YP_010894795.1 | abs/path/to/protein_references/BSNHM038-24.fasta | BSNHM038-24 |
| BSNHM046-24 | Polycentropodidae | YP_010426350.1 | abs/path/to/protein_references/BSNHM046-24.fasta | BSNHM046-24 |
  
  

## Running: ##
**snakefile:** 
- Uses config-genefetch.yaml
  - Contains path to 'samples.csv', an output directory, and 'protein_references.csv'
  - Contains different parameter configurations for -s and -r
- Uses taxa-specific references from protein_references.csv (requires 1-gene_fetch.py to be run (test data protein_references.csv example provided)).
- 3_mge_tidy-snakemake.py and 4_mge_stats-snakemake.py functionality integrated into snakefile

**snakefile-fastpmerge:** 
- Utilises a fastp merge and post-merge concatenation of merged and unpaired reads as an alternative to the 'standard' MGE pipeline.

**Test run**
- Raw reads for 12 test samples can be downloaded [here](https://naturalhistorymuseum-my.sharepoint.com/personal/b_price_nhm_ac_uk/_layouts/15/onedrive.aspx?ct=1723035606962&or=Teams%2DHL&ga=1&LOF=1&id=%2Fpersonal%2Fb%5Fprice%5Fnhm%5Fac%5Fuk%2FDocuments%2F%5Ftemp%2F%5FBGEexamples4Felix%2F1%5Fraw%5Fdata). Each read pair must be in seperate subdirectories under a parent directory that can be called anything
- samples sheet (BGE_test_samples.csv) provided (paths to reads and references need to be altered)
- protein_references sheet (gene_fetch_BGE_test_data.csv) provided (in protein_references/).


## To do ##
- Integrate 1_gene_fetch.py into snakefile
- Make Workflow Hub compatible
- Generate RO-crates
  
## MGE output post-processing ##
As an alternative to running 'snakefile' or 'snakefile_genefetch-fastpmerge', 3_mge_tidy-snakemake.py and 4_mge_stats-snakemake.py supplementary scripts can be run seperately after MGE.

Python scripts to organise outputs into directories, trim consensus sequences and 'clean' header names, concatenate into a multi-fasta, calculate summary stats and run BLAST on each consensus sequence.

### 3_mge_tidy-snakemake.py ###
Usage information output by running 'python 3_mge_tidy-snakemake.py'
- Organised files into subdirectories (alignment, consensus, err, logs, out) depending on file extension.
- Concatenates consensus.fas files together (within consensus subdir) and outputs user-specified multi-fasta (highly recommended to include 'cox1' in concatenated consensus filename for downstream BOLDigger2 runs).
- 'Cleans' and trims sequence headers of each consenus sequences.

### 4_mge_stats-snakemake.py ###
Usage information output by running 'python 4_mge_stats-snakemake.py'. Outputs summary stats to .csv file.
- Reads relevant fields from .out files in 'out' dir:
  -  consensus sequence output (i.e. process ID)
  -  Number of input sequences considered
  -  sequences found in vulgar file
  -  number of aligned reads
  -  number of skipped reads due to low rel. score (low relative score = read didn't pass exonerate relative score threshold (i.e. -r))
  -  length of alignment (i.e. reference sequence length)
  -  Coverage minimum
  -  Coverage maximum
  -  Coverage mean
  -  Coverage median
- Determines sequence stats from concatenated consensus.fasta generated by '4_mge_tidy-snakemake.py':
  - Sequence length (only GTACs)
  - N Count
  - Dash and Tilde Count
  - Non-GTAC Count (total of 'N Count' and 'Dash and Tilde Count')
